### PR TITLE
Rework configuration

### PR DESCRIPTION
**Pull request by AlanGriffiths**
There are just three snap configuration options used:

1. daemon=[true|false] run the daemon
2. kiosk-config=\<contents for ubuntu-core-kiosk.config>
3. kiosk-display=\<contents for ubuntu-core-kiosk.display>

This is both simpler and more flexible than the approach that evolved for mir-kiosk.

### CODE REVIEW

File: snap/hooks/configure [hunk 1]
```diff
@@ -1,158 +1,90 @@
...
-if [ ! -e "${kiosk_config}" ]; then
-cat <<EOT > "${kiosk_config}"
-arw-file=
-console-provider=vt
-EOT
-let config_changes+=1
-fi
+config_entry() {
+  if ! grep -q "^[# ]*${1##?[#\ ]}=" "${config_file}"; then
+    cat <<EOT >> "${config_file}"
```
Comment (Saviq) [inline snap/hooks/configure]: If this is only used for the defaults, couldn't we just `<<EOF` the whole thing in one go?

File: snap/hooks/configure [hunk 2]
```diff
@@ -1,158 +1,90 @@
...
+      mv "${display_file}~" "${display_file}"
+      let config_changes+=1
     else
-      sed --in-place=.save '/^vt/d' ${kiosk_config}
+      rm "${display_file}~"
     fi
-    echo vt=${vt} >> ${kiosk_config}
-    let config_changes+=1
   fi
 fi
```
Comment (Saviq) [inline snap/hooks/configure]: You could have a common function taking the config key, file path and default contents to deal with both the display and config?

File: snap/hooks/configure [hunk 3]
```diff
@@ -1,158 +1,90 @@
...
     else
-      sed --in-place=.save '/^cursor/d' ${kiosk_config}
+      rm "${config_file}~"
     fi
-    echo cursor=${cursor} >> ${kiosk_config}
-    let config_changes+=1
-  fi
-fi
+  else
+    # No kiosk configuration is supplied, we'll generate a default
```
Comment (Saviq) [inline snap/hooks/configure]: Should we `snapctl set kiosk-config` to the default, then?

Or is the idea that until the user `snap set`s, we'll always regenerate the defaults?

File: snap/hooks/configure [hunk 4]
```diff
@@ -1,158 +1,90 @@
...
-  elif [ $(($config_changes+$display_changes)) -gt 0 ]; then
-    snapctl restart $SNAP_NAME || true
+  if snapctl services "${SNAP_NAME}.daemon" | grep -q inactive; then
+    snapctl start --enable "$SNAP_NAME.daemon" 2>&1 || true
+  elif [ $config_changes -gt 0 ]; then
+    snapctl restart "$SNAP_NAME.daemon" || true
   fi
 else
-  snapctl stop --disable $SNAP_NAME.daemon 2>&1 || true
+  snapctl stop --disable "$SNAP_NAME.daemon" 2>&1 || true
```
Comment (Saviq) [inline snap/hooks/configure]: `s/SNAP_NAME/SNAP_INSTANCE_NAME/`

**Review by Saviq — CHANGES_REQUESTED** [review]
`post-refresh` could just be a symlink to `install`. That said, do we need/want it? It was there for transitioning IIRC?

I'd go for `set -euo pipefail` across the board. Maybe even `-x`, too.

In general I see the appeal of this approach, but I think it needs more work. How does a user only set a single key? If the first value in the config wins, could we put the user's config first, then our defaults? Or the other way around?

How do we deal with, inevitable, transitions?

File: snap/hooks/configure (hunk 1 above)
Comment (AlanGriffiths) [reply snap/hooks/configure]: Your premise is false: If the file exists, but comes from a different version (or has been edited) we want to fill in any missing options.

File: snap/hooks/configure (hunk 3 above)
Comment (AlanGriffiths) [reply snap/hooks/configure]: The idea is that we merge the options we need with those in the config unless overridden by a snap option.

**Comment by AlanGriffiths** [comment]
> `post-refresh` could just be a symlink to `install`. That said, do we need/want it? It was there for transitioning IIRC?

Yes, and it is needed for transitions. Remember, a user can switch from `edge` to `stable`, so we can't assume that there isn't another version yet.

> I'd go for `set -euo pipefail` across the board. Maybe even `-x`, too.

Ack.

> In general I see the appeal of this approach, but I think it needs more work. How does a user only set a single key? If the first value in the config wins, could we put the user's config first, then our defaults? Or the other way around?

I'm not sure how boost.Options handles duplicate keys, but I don't want to confuse users (nor rely on non-obvious behaviour).

> How do we deal with, inevitable, transitions?

By having `post-refresh`?

Thanks for the feedback! I'll iterate.

File: snap/hooks/configure (hunk 1 above)
Comment (Saviq) [reply snap/hooks/configure]: Why don't we apply the same "DO NOT EDIT" approach to the file, then?

File: snap/hooks/configure (hunk 3 above)
Comment (Saviq) [reply snap/hooks/configure]: This feels fragile. I'd rather the config file was owned by the configure hook.

**Comment by Saviq** [comment]
> Yes, and it is needed for transitions. Remember, a user can switch from `edge` to `stable`, so we can't assume that there isn't another version yet.

Well, I don't think we need to care for that until we release outside of `edge` ;)

> I'm not sure how boost.Options handles duplicate keys, but I don't want to confuse users (nor rely on non-obvious behaviour).

That would be an internal detail, though, if we own the config file. But it would simplify the story of _having_ defaults in the first place.

File: snap/hooks/configure (hunk 1 above)
Comment (AlanGriffiths) [reply snap/hooks/configure]: I assumed editing is the normal and easiest way to try things out, or set up a one-off system. But the flow could be:

1. Write the `kiosk-config` value into a temporary file
2. merge with the defaults (same as proposed logic, but using the temporary file)
3. diff with the current file
4. Profit

WDYT?

Comment (Saviq) [reply snap/hooks/configure]: Yeah, I think if we give a `snap set` way, allowing for direct file editing would be bound to cause grief.

I would still be happier if we could just `echo $DEFAULTS $USER_CONFIG > config.file`. Can you please find out if Boost.Options has defined behaviour with duplicate keys? It's customary to respect the last entry, hoping that would be the case here.

File: snap/hooks/configure [hunk 5]
```diff
@@ -1,158 +1,84 @@
...
-      sed --in-place=.save '/^cursor/d' ${kiosk_config}
-    fi
-    echo cursor=${cursor} >> ${kiosk_config}
-    let config_changes+=1
-  fi
+# WORKAROUND: If 'vt.handoff' is set, Mir can only start with vt=0.
+# The correct fix is: https://github.com/snapcore/pi-gadget/pull/12
+vt=0
+EOT
 fi
```
Comment (Saviq) [inline snap/hooks/configure]: Is `vt.handoff` alone that triggers this? Should you check for `snap_core` then?

File: snap/hooks/configure [hunk 6]
```diff
@@ -1,158 +1,84 @@
...
+  rm "${config_temp}"
 fi
 
+# (Re)start the daemon
 if [ "$(snapctl get daemon)" = "true" ]; then
-  if snapctl services ${SNAP_NAME}.daemon | grep -q inactive; then
-    snapctl start --enable $SNAP_NAME.daemon 2>&1 || true
-  elif [ $(($config_changes+$display_changes)) -gt 0 ]; then
-    snapctl restart $SNAP_NAME || true
+  if snapctl services "${SNAP_INSTANCE_NAME}.daemon" | grep -q inactive; then
```
Comment (Saviq) [inline snap/hooks/configure]:
```suggestion
  if snapctl services "$SNAP_INSTANCE_NAME.daemon" | grep -q inactive; then
```

For consistency.

**Review by Saviq — COMMENTED** [review]
Still think we can do away with `post-refresh` for now.

File: snap/hooks/configure (hunk 5 above)
Comment (AlanGriffiths) [reply snap/hooks/configure]: That's the logic in mir-kiosk, I'll dig to find out how we got there.

**Comment by AlanGriffiths** [comment]
> Still think we can do away with `post-refresh` for now.

I disagree. It is needed for any version being switched to. And it is possible for a user to switch to an older version. For example, by switching channels or by `snap revert`.

**Comment by AlanGriffiths** [comment]
> I would still be happier if we could just echo $DEFAULTS $USER_CONFIG > config.file. Can you please find out if Boost.Options has defined behaviour with duplicate keys? It's customary to respect the last entry, hoping that would be the case here.

I still think that leads to a confusing config file. Even if users are not expected to edit the file directly, they can (and should be able to) read it.

In any case, I can't find explicit documentation about how duplicate keys are handled from a config file, but the implication is that it takes the first value, and empirically, that is what happens.

**Pull request merged**